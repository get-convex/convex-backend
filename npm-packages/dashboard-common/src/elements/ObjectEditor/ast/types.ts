// eslint-disable-next-line max-classes-per-file
import { Node as AcornNode, Position, SourceLocation } from "acorn";
import { ValidatorJSON, Value } from "convex/values";
import type { editor } from "monaco-editor";
import { displayValidatorPretty } from "../../../lib/format";
import { typeForValue } from "./helpers";

export type Node =
  | ProgramNode
  | ExpressionNode
  | NewExpressionNode
  | ObjectNode
  | ObjectPropertyNode
  | ArrayNode
  | UnaryExpressionNode
  | IdentifierNode
  | CallExpressionNode
  | LiteralNode
  | TemplateLiteralNode;

// Acorn doesn't fully type nested nodes so we do it ourselves.
export type ProgramNode = {
  type: "Program";
  body: [Node];
} & AcornNode;

export type ExpressionNode = {
  type: "ExpressionStatement";
  expression: Node;
} & AcornNode;

export type NewExpressionNode = {
  type: "NewExpression";
  callee: IdentifierNode;
  arguments: LiteralNode[];
} & AcornNode;

export type ObjectNode = {
  type: "ObjectExpression";
  properties: ObjectPropertyNode[];
} & AcornNode;

export type ObjectPropertyNode = {
  type: "Property";
  key: IdentifierNode | LiteralNode;
  value: Node;
} & AcornNode;

export type ArrayNode = {
  type: "ArrayExpression";
  elements: (Node | null)[];
} & AcornNode;

export type UnaryExpressionNode = {
  type: "UnaryExpression";
  operator: string;
  prefix: boolean;
  argument: Node;
} & AcornNode;

export type IdentifierNode = { type: "Identifier"; name: string } & AcornNode;
export type CallExpressionNode = {
  type: "CallExpression";
  callee: { name?: string };
  arguments: Node[];
} & AcornNode;

export type LiteralNode = {
  type: "Literal";
  value: any;
  regex?: any;
} & AcornNode;

export type TemplateElementNode = {
  type: "TemplateElement";
  value: { cooked: string; raw: string };
  tail: boolean;
} & AcornNode;

export type TemplateLiteralNode = {
  type: "TemplateLiteral";
  quasis: TemplateElementNode[];
  expressions: Node[];
} & AcornNode;

export class ConvexValidationError extends Error {
  public readonly markerData: Partial<editor.IMarkerData>;

  public readonly loc?: SourceLocation;

  constructor(
    message: string,
    loc?: SourceLocation,
    markerData?: Partial<editor.IMarkerData>,
  ) {
    super(message);
    this.name = "ConvexValidationError";
    this.markerData = markerData ?? {};
    // Because we're adding an extra character at the start of the first line, the
    // left open paren, ConvexValidationError has to shift any location on the first
    // line back by one.
    if (loc) {
      this.loc = {
        ...loc,
        start: fixPosition(loc.start),
        end: fixPosition(loc.end),
      };
    }
  }
}

export type SchemaValidationErrorCode =
  | "IsNotArray"
  | "IsNotObject"
  | "IsNotBytes"
  | "RecordKeysMismatch"
  | "ObjectMismatch"
  | "UnionMismatch"
  | "LiteralMismatch"
  | "ExtraProperty"
  | "RequiredPropertyMissing";
export class ConvexSchemaValidationError extends ConvexValidationError {
  public readonly code: SchemaValidationErrorCode;

  constructor(
    code: SchemaValidationErrorCode,
    validator: ValidatorJSON,
    value?: Value,
    loc?: SourceLocation,
    markerData?: Partial<editor.IMarkerData>,
  ) {
    const message =
      code === "RequiredPropertyMissing"
        ? `${value ? `Property '${value}'` : "Value"} is missing but required in schema:\n${displayValidatorPretty(validator)}`
        : code === "ExtraProperty"
          ? `Property '${value}' does not exist in schema:\n${displayValidatorPretty(validator)}`
          : code === "UnionMismatch"
            ? `Object does not match any of the types in union:\n${displayValidatorPretty(validator)}`
            : `Type '${value !== undefined ? typeForValue(value) : "Value"}' is not assignable to:\n${displayValidatorPretty(validator)}`;
    super(message, loc, markerData);
    this.name = "ConvexSchemaValidationError";
    this.code = code;
  }
}

function fixPosition(position: Position) {
  return position.line === 1
    ? { ...position, column: position.column - 1 }
    : position;
}

export type WalkResults = {
  value: Value;
  errors: ConvexValidationError[];
};

export const isUndefined = (node: Node) =>
  node.type === "Identifier" && node.name === "undefined";
