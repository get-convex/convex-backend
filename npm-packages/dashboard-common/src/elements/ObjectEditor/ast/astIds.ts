import { isId } from "id-encoding";
import { Node, LiteralNode } from "./types";

export class IdWalker {
  private ids: LiteralNode[] = [];

  walk(n: Node): LiteralNode[] {
    switch (n.type) {
      case "ArrayExpression": {
        n.elements.forEach((element) => {
          if (element !== null) {
            this.walk(element);
          }
        });
        break;
      }

      case "ObjectExpression": {
        n.properties.forEach((property) => {
          this.walk(property.key);
          this.walk(property.value);
        });
        break;
      }

      case "UnaryExpression": {
        break;
      }

      case "NewExpression": {
        break;
      }

      case "TemplateLiteral": {
        n.expressions?.forEach((expr) => this.walk(expr));
        break;
      }

      case "Literal": {
        if (typeof n.value === "string" && isId(n.value)) {
          this.ids.push(n);
        }
        break;
      }

      case "Identifier": {
        // Identifiers are not string literals, so we skip them.
        break;
      }

      case "CallExpression": {
        break;
      }

      default: {
        // Unsupported syntax, do nothing.
        break;
      }
    }

    return this.ids;
  }
}
