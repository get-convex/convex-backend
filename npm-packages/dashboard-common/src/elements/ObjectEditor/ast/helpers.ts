import { ValidatorJSON, Value } from "convex/values";
import isPlainObject from "lodash/isPlainObject";
import { UNDEFINED_PLACEHOLDER } from "system-udfs/convex/_system/frontend/patchDocumentsFields";
import * as IdEncoding from "id-encoding";

export const isValidValue = (
  validator: ValidatorJSON,
  value: Value,
  shallow = true,
): boolean => {
  switch (validator.type) {
    case "null":
      return value === null;
    case "string":
      return typeof value === "string";
    case "boolean":
      return typeof value === "boolean";
    case "number":
      return typeof value === "number";
    case "bigint":
      return typeof value === "bigint";
    case "bytes":
      return value instanceof Uint8Array;
    case "any":
      return true;
    case "literal":
      return value === validator.value && value !== UNDEFINED_PLACEHOLDER;
    case "id":
      // TODO: Validate table names
      return typeof value === "string" && IdEncoding.isId(value);
    case "object": {
      const isObject =
        typeof value === "object" && value !== null && isPlainObject(value);
      if (shallow || !isObject) {
        return isObject;
      }
      for (const [key, valueValidator] of Object.entries(validator.value)) {
        const nextValue = (value as Record<string, Value>)[key];
        if (valueValidator.optional && nextValue === undefined) {
          continue;
        }

        if (!isValidValue(valueValidator.fieldType, nextValue, shallow)) {
          return false;
        }
      }
      return true;
    }
    case "record": {
      const isObject =
        typeof value === "object" && value !== null && isPlainObject(value);
      if (shallow || !isObject) {
        return isObject;
      }
      for (const [key, keyValue] of Object.entries(value)) {
        if (
          !isValidValue(validator.keys, key, shallow) ||
          !isValidValue(validator.values.fieldType, keyValue, shallow)
        ) {
          return false;
        }
      }
      return true;
    }
    case "array": {
      const isArray = Array.isArray(value);
      if (shallow || !isArray) {
        return isArray;
      }

      for (const v of value) {
        if (!isValidValue(validator.value, v, shallow)) {
          return false;
        }
      }

      return true;
    }
    case "union":
      return validator.value.some((v) =>
        // This is still shallow because we need to check every type in the union.
        isValidValue(v, value, shallow),
      );
    default: {
      const _typeCheck: never = validator;
      return false;
    }
  }
};

export const typeForValue = (value: Value): ValidatorJSON["type"] => {
  if (value === null) {
    return "null";
  }
  if (typeof value === "string") {
    return "string";
  }
  if (typeof value === "boolean") {
    return "boolean";
  }
  if (typeof value === "number") {
    return "number";
  }
  if (typeof value === "bigint") {
    return "bigint";
  }
  if (value instanceof Uint8Array) {
    return "bytes";
  }
  if (isPlainObject(value)) {
    return "object";
  }
  if (Array.isArray(value)) {
    return "array";
  }
  return "any";
};
