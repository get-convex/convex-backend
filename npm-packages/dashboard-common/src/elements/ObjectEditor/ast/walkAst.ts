import { Parser } from "acorn";
import { ValidatorJSON } from "convex/values";
import { UNDEFINED_PLACEHOLDER } from "system-udfs/convex/_system/frontend/patchDocumentsFields";
import { Node, ExpressionNode, isUndefined } from "./types";
import { walkPatchDocuments } from "./walkPatchDocuments";
import { Walker } from "./ast";
import { walkMultipleDocuments } from "./walkMultipleDocuments";
import { IdWalker } from "./astIds";

export type WalkAstOptions = {
  validator?: ValidatorJSON;
} & (
  | { mode: "addDocuments" | "editDocument" | "patchDocuments" }
  | {
      mode: "editField";
      allowTopLevelUndefined?: boolean;
    }
);

export const walkAst = (code: string, options: WalkAstOptions) => {
  const ast = Parser.parse(`(${code})`, {
    ecmaVersion: 2020,
    locations: true,
  }) as Node;

  // Peel off the unneeded nodes off the AST.
  const programBody = extractExpressionFromAST(ast);

  const { value, errors } = extractValueFromExpression(programBody, options);

  const idFinder = new IdWalker();

  const ids = idFinder.walk(programBody.expression);

  return {
    value,
    errors,
    ids,
  };
};

function extractExpressionFromAST(ast: Node) {
  if (ast.type !== "Program" || ast.body.length !== 1) {
    throw new Error("Invalid AST, expected Program with one node in body");
  }

  const programBody = ast.body[0];
  if (programBody.type !== "ExpressionStatement") {
    throw new Error("Invalid AST, expected ExpressionStatement");
  }
  return programBody;
}

function extractValueFromExpression(
  programBody: ExpressionNode,
  options: WalkAstOptions,
) {
  const { mode, validator } = options;
  const { expression } = programBody;

  if (
    mode === "editField" &&
    options.allowTopLevelUndefined &&
    isUndefined(expression)
  ) {
    return { value: UNDEFINED_PLACEHOLDER, errors: [] };
  }

  if (mode === "patchDocuments") {
    return walkPatchDocuments(expression, validator);
  }

  if (mode === "addDocuments" && expression.type === "ArrayExpression") {
    return walkMultipleDocuments(expression, validator);
  }

  // We are not editing multiple documents, so we can walk as a single node.
  const walker = new Walker({ validator });

  const isTopLevel = mode === "editDocument" || mode === "addDocuments";
  return walker.walk(expression, isTopLevel);
}
