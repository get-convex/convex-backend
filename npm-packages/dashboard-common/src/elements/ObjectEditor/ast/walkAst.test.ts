import { ValidatorJSON, Value } from "convex/values";
import { UNDEFINED_PLACEHOLDER } from "system-udfs/convex/_system/frontend/patchDocumentsFields";
import * as Base64 from "base64-js";
import { stringifyValue } from "../../../lib/stringifyValue";
import { walkAst } from "./walkAst";
import { SchemaValidationErrorCode } from "./types";
import { validateConvexFieldName } from "./ast";

const validators: Record<string, ValidatorJSON> = {
  number: { type: "number" },
  bigint: { type: "bigint" },
  boolean: { type: "boolean" },
  string: { type: "string" },
  null: { type: "null" },
  id: { type: "id", tableName: "myTable" },
  any: { type: "any" },
};

const unionOfObjectValidator: ValidatorJSON = {
  type: "union",
  value: [
    {
      type: "object",
      value: {
        kind: {
          fieldType: { type: "literal", value: "TEXT" },
          optional: false,
        },
        text: { fieldType: validators.string, optional: false },
      },
    },
    {
      type: "object",
      value: {
        kind: {
          fieldType: { type: "literal", value: "IMAGE" },
          optional: false,
        },
        uri: { fieldType: validators.string, optional: false },
      },
    },
  ],
};

const unionOfArrayValidator: ValidatorJSON = {
  type: "union",
  value: [
    {
      type: "array",
      value: validators.number,
    },
    {
      type: "array",
      value: validators.string,
    },
  ],
};

describe("walkAst", () => {
  describe("happy path", () => {
    const testHappyPath = (
      input: string,
      expected: Value,
      mode: "editField" | "addDocuments" | "editDocument" | "patchDocuments",
      allowTopLevelUndefined: boolean,
      validator?: ValidatorJSON,
    ) => {
      const { value: result, errors } = walkAst(input, {
        mode,
        allowTopLevelUndefined,
        validator,
      });
      expect(errors).toEqual([]);
      expect(result).toEqual(expected);

      // ensure that parsing the string version of a value produces the same result.
      const { value: roundTripResult, errors: roundTripErrors } = walkAst(
        stringifyValue(result),
        {
          mode,
        },
      );
      expect(roundTripResult).toEqual(result);
      expect(roundTripErrors).toEqual([]);
    };

    const testCases: {
      name: string;
      input: string;
      expected: Value;
      mode?: "editField" | "addDocuments" | "editDocument" | "patchDocuments";
      allowTopLevelUndefined?: boolean;
      validator?: ValidatorJSON;
    }[] = [
      {
        name: "wrapped in parens",
        input: "(({a: 2}))",
        expected: { a: 2 },
      },
      {
        name: "number",
        input: "1.25",
        expected: 1.25,
        validator: validators.number,
      },
      {
        name: "infinity",
        input: "Infinity",
        expected: Infinity,
        validator: validators.number,
      },
      {
        name: "infinity",
        input: "Infinity",
        expected: Infinity,
        validator: validators.number,
      },
      {
        name: "-infinity",
        input: "-Infinity",
        expected: -Infinity,
        validator: validators.number,
      },
      {
        name: "nan",
        input: "NaN",
        expected: NaN,
        validator: validators.number,
      },
      {
        name: "negative number",
        input: "-1.25",
        expected: -1.25,
        validator: validators.number,
      },
      {
        name: "bigint",
        input: "3n",
        expected: BigInt(3),
        validator: validators.bigint,
      },
      {
        name: "negative bigint",
        input: "-3n",
        expected: BigInt(-3),
        validator: validators.bigint,
      },
      {
        name: "boolean",
        input: "true",
        expected: true,
        validator: validators.boolean,
      },
      {
        name: "string",
        input: `"a string"`,
        expected: "a string",
        validator: validators.string,
      },
      {
        name: "single-quote string",
        input: "'a string'",
        expected: "a string",
        validator: validators.string,
      },
      {
        name: "null",
        input: "null",
        expected: null,
        validator: validators.null,
      },
      {
        name: "array",
        input: `[1,"abc",false]`,
        expected: [1, "abc", false],
        validator: {
          type: "array",
          value: {
            type: "union",
            value: [validators.number, validators.string, validators.boolean],
          },
        },
      },
      {
        name: "object",
        input: `{a: 1}`,
        expected: { a: 1 },
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.number, optional: false },
          },
        },
      },
      {
        name: "template literal",
        input: "`abc\ndef`",
        expected: "abc\ndef",
        validator: validators.string,
      },
      {
        name: "object with literal keys",
        input: `{"a": 1}`,
        expected: { a: 1 },
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.number, optional: false },
          },
        },
      },
      {
        name: "object with empty key",
        input: `{"": 1}`,
        expected: { "": 1 },
        validator: {
          type: "object",
          value: {
            "": { fieldType: validators.number, optional: false },
          },
        },
      },
      {
        name: "nested object",
        input: "{a: {b: 2}}",
        expected: { a: { b: 2 } },
        validator: {
          type: "object",
          value: {
            a: {
              fieldType: {
                type: "object",
                value: {
                  b: { fieldType: validators.number, optional: false },
                },
              },
              optional: false,
            },
          },
        },
      },
      {
        name: "nested object with system field",
        input: "{a: {_a: 1}}",
        expected: { a: { _a: 1 } },
        validator: {
          type: "object",
          value: {
            a: {
              fieldType: {
                type: "object",
                value: {
                  _a: { fieldType: validators.number, optional: false },
                },
              },
              optional: false,
            },
          },
        },
      },
      {
        name: "object with nested array",
        input: "{a: [1,2,3]}",
        expected: { a: [1, 2, 3] },
        validator: {
          type: "object",
          value: {
            a: {
              fieldType: {
                type: "array",
                value: validators.number,
              },
              optional: false,
            },
          },
        },
      },
      {
        name: "array with nested object",
        input: "[{a:1}]",
        expected: [{ a: 1 }],
        validator: {
          type: "array",
          value: {
            type: "object",
            value: {
              a: { fieldType: validators.number, optional: false },
            },
          },
        },
      },
      {
        name: "id",
        input: "'2wbatrp2sqy0ym0j6xra4agx9h059j0'",
        expected: "2wbatrp2sqy0ym0j6xra4agx9h059j0",
        validator: validators.id,
      },
      {
        name: "object with nested id",
        input: "{a: '2wbatrp2sqy0ym0j6xra4agx9h059j0'}",
        expected: { a: "2wbatrp2sqy0ym0j6xra4agx9h059j0" },
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.id, optional: false },
          },
        },
      },
      {
        name: "object with nested ids all over the place",
        input:
          "{a: '2wbatrp2sqy0ym0j6xra4agx9h059j0', b: ['31edfed94p29qg8byf54zm1s9h05xtg', '2z16se66x9g9jsgd759nb3fw9h00ymr']}",
        expected: {
          a: "2wbatrp2sqy0ym0j6xra4agx9h059j0",
          b: [
            "31edfed94p29qg8byf54zm1s9h05xtg",
            "2z16se66x9g9jsgd759nb3fw9h00ymr",
          ],
        },
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.id, optional: false },
            b: {
              fieldType: {
                type: "array",
                value: validators.id,
              },
              optional: false,
            },
          },
        },
      },
      {
        name: "top level undefined",
        allowTopLevelUndefined: true,
        input: "undefined",
        expected: UNDEFINED_PLACEHOLDER,
        // Even though the value is undefined, allowTopLevelUndefined is true, so the validator should bei gnored.
        validator: validators.number,
      },
      {
        name: "top level undefined in object",
        allowTopLevelUndefined: true,
        mode: "patchDocuments",
        input: "{a: undefined}",
        expected: { a: UNDEFINED_PLACEHOLDER },
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.number, optional: true },
          },
        },
      },
      {
        name: "top level union in object",
        mode: "patchDocuments",
        input: "{a: 123}",
        expected: { a: 123 },
        validator: {
          type: "union",
          value: [
            {
              type: "object",
              value: {
                a: { fieldType: validators.number, optional: false },
                b: { fieldType: validators.string, optional: false },
              },
            },
          ],
        },
      },
      {
        name: "undefined in union",
        input: "{}",
        expected: {},
        validator: {
          type: "union",
          value: [
            {
              type: "object",
              value: {
                a: { fieldType: validators.number, optional: true },
              },
            },
          ],
        },
      },
      {
        name: "multiline string with backticks",
        input: "`abc\\`\ndef`",
        expected: "abc`\ndef",
        validator: validators.string,
      },
      {
        name: "union of objects",
        input: `{kind: "TEXT", text: "abc"}`,
        expected: { kind: "TEXT", text: "abc" },
        validator: unionOfObjectValidator,
      },
      {
        name: "union of objects 2",
        input: `{kind: "IMAGE", uri: "abc"}`,
        expected: { kind: "IMAGE", uri: "abc" },
        validator: unionOfObjectValidator,
      },
      {
        name: "union of arrays",
        input: `[1,2,3]`,
        expected: [1, 2, 3],
        validator: unionOfArrayValidator,
      },
      {
        name: "union of arrays 2",
        input: `["a","b","c"]`,
        expected: ["a", "b", "c"],
        validator: unionOfArrayValidator,
      },
      {
        name: "object with any",
        input: `{a: 1, b: "abc"}`,
        expected: { a: 1, b: "abc" },
        validator: validators.any,
      },
      {
        name: "object with any key",
        input: `{a: 1, b: "abc"}`,
        expected: { a: 1, b: "abc" },
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.any, optional: false },
            b: { fieldType: validators.any, optional: false },
          },
        },
      },
      {
        name: "array with any",
        input: `[1, "abc"]`,
        expected: [1, "abc"],
        validator: validators.any,
      },
      {
        name: "array with any value",
        input: `[1, "abc"]`,
        expected: [1, "abc"],
        validator: { type: "array", value: validators.any },
      },
      {
        name: "any primitive",
        input: "1",
        expected: 1,
        validator: validators.any,
      },
      {
        name: "record validator",
        input: `{a: 1, b: "abc"}`,
        expected: { a: 1, b: "abc" },
        validator: {
          type: "record",
          keys: { type: "string" },
          values: {
            fieldType: {
              type: "union",
              value: [validators.number, validators.string],
            },
            optional: false,
          },
        },
      },
      {
        name: "record validator with union of ids",
        input: `{j9728bhc7wsqs6aptq2j6p4e496z8gbr: 1}`,
        expected: { j9728bhc7wsqs6aptq2j6p4e496z8gbr: 1 },
        validator: {
          type: "record",
          keys: {
            type: "union",
            value: [
              { type: "id", tableName: "abc" },
              { type: "id", tableName: "abc2" },
            ],
          },
          values: {
            fieldType: {
              type: "union",
              value: [validators.number, validators.string],
            },
            optional: false,
          },
        },
      },
    ];

    test.each(testCases)(
      "$name",
      ({
        input,
        expected,
        allowTopLevelUndefined = false,
        mode = "editField",
        validator = undefined,
      }) => {
        // Test with and without a validator.
        testHappyPath(input, expected, mode, allowTopLevelUndefined, validator);
        testHappyPath(input, expected, mode, allowTopLevelUndefined, undefined);
      },
    );
  });

  describe("error cases", () => {
    test.each([
      { name: "regex", input: "/a/" },
      { name: "invalid unary expression", input: "+3" },
      { name: "negative string", input: `-"a"` },
      { name: "negative object", input: `-{}` },
      { name: "unsupported constructor", input: "new User()" },
      { name: "Id class", input: "new Id()" },
      { name: "key starting with $", input: '{"$a": 1}' },
      { name: "non-identifier key", input: '{"foo-bar": 1}' },
      { name: "identifier", input: "a" },
      { name: "identifier in object", input: "{a: b}" },
      { name: "array with empty value", input: "[1,,3]" },
      // eslint-disable-next-line no-template-curly-in-string
      { name: "template literal with expression", input: "`${'abc'}`" },
      {
        name: "top level undefined",
        input: "undefined",
      },
      {
        name: "top level undefined in object",
        input: "{a: undefined}",
      },
      {
        name: "nested undefined in object even if top level is allowed",
        input: "{a: {b: undefined}}",
        allowTopLevelUndefined: true,
      },
      {
        name: "CallExpression that isn't bytes",
        input: "foo()",
      },
      {
        name: "Bytes with no arguments",
        input: "Bytes()",
      },
      {
        name: "Bytes with too many arguments",
        input: `Bytes("a", "a")`,
      },
      {
        name: "Bytes with bad argument",
        input: `Bytes("aaa")`,
      },
      { name: "-nan", input: "-NaN" },
      { name: "top level undefined in object", input: "{a: undefined}" },
    ])("$name", ({ input, allowTopLevelUndefined = false }) => {
      const { errors } = walkAst(input, {
        mode: "editField",
        allowTopLevelUndefined,
      });

      expect(errors).toMatchSnapshot();
    });
  });

  describe("top level system field disallowed", () => {
    const { errors } = walkAst('[{"_a": 1}]', { mode: "addDocuments" });
    expect(errors.length).toBeGreaterThan(0);
    expect(errors).toMatchSnapshot();
  });

  describe("validator error cases", () => {
    const testCases: {
      name: string;
      input: string;
      validator: ValidatorJSON;
      errorCode: SchemaValidationErrorCode;
      expectedValue?: Value;
    }[] = [
      {
        name: "Top-level literal mismatch",
        input: "'string'",
        expectedValue: "string",
        errorCode: "LiteralMismatch",
        validator: validators.number,
      },
      {
        name: "Top-level template literal",
        input: "`abc`",
        expectedValue: "abc",
        errorCode: "LiteralMismatch",
        validator: validators.number,
      },
      {
        name: "Object missing property",
        input: "{}",
        expectedValue: {},
        errorCode: "RequiredPropertyMissing",
        validator: {
          type: "object",
          value: {
            a: {
              fieldType: validators.number,
              optional: false,
            },
          },
        },
      },
      {
        name: "Extra property",
        input: "{a: 1, b: 2}",
        expectedValue: { a: 1, b: 2 },
        errorCode: "ExtraProperty",
        validator: {
          type: "object",
          value: { a: { fieldType: validators.number, optional: false } },
        },
      },
      {
        name: "Input is array, but validator is not",
        input: "[]",
        expectedValue: [],
        errorCode: "IsNotArray",
        validator: validators.string,
      },
      {
        name: "Input is object, but validator is not",
        input: "{}",
        expectedValue: {},
        errorCode: "IsNotObject",
        validator: validators.string,
      },
      {
        name: "Input is bytes, but validator is not",
        input: "Bytes('aaaa')",
        expectedValue: Base64.toByteArray("aaaa").buffer,
        errorCode: "IsNotBytes",
        validator: validators.string,
      },
      {
        name: "Literal in object",
        input: "{a: 1}",
        expectedValue: { a: 1 },
        errorCode: "LiteralMismatch",
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.string, optional: false },
          },
        },
      },
      {
        name: "Array with invalid element",
        input: "[1, 'abc']",
        expectedValue: [1, "abc"],
        errorCode: "LiteralMismatch",
        validator: {
          type: "array",
          value: validators.number,
        },
      },
      {
        name: "Object with invalid property value",
        input: "{ a: 'abc' }",
        expectedValue: { a: "abc" },
        errorCode: "LiteralMismatch",
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.number, optional: false },
          },
        },
      },
      {
        name: "Nested object with invalid property value",
        input: "{ a: { b: 'abc' } }",
        expectedValue: { a: { b: "abc" } },
        errorCode: "LiteralMismatch",
        validator: {
          type: "object",
          value: {
            a: {
              fieldType: {
                type: "object",
                value: {
                  b: { fieldType: validators.number, optional: false },
                },
              },
              optional: false,
            },
          },
        },
      },
      {
        name: "Array with invalid nested object",
        input: "[{ a: 'abc' }]",
        expectedValue: [{ a: "abc" }],
        errorCode: "LiteralMismatch",
        validator: {
          type: "array",
          value: {
            type: "object",
            value: {
              a: { fieldType: validators.number, optional: false },
            },
          },
        },
      },
      {
        name: "Object with invalid nested array",
        input: "{ a: ['abc', 123] }",
        expectedValue: { a: ["abc", 123] },
        errorCode: "LiteralMismatch",
        validator: {
          type: "object",
          value: {
            a: {
              fieldType: {
                type: "array",
                value: validators.number,
              },
              optional: false,
            },
          },
        },
      },
      {
        name: "Object with invalid nested id",
        input: "{ a: 'invalid-id' }",
        expectedValue: { a: "invalid-id" },
        errorCode: "LiteralMismatch",
        validator: {
          type: "object",
          value: {
            a: { fieldType: validators.id, optional: false },
          },
        },
      },
      {
        name: "union of objects",
        input: `{a: 1}`,
        expectedValue: { a: 1 },
        errorCode: "UnionMismatch",
        validator: unionOfObjectValidator,
      },
      {
        name: "union of arrays",
        input: `{a: 1}`,
        expectedValue: { a: 1 },
        errorCode: "UnionMismatch",
        validator: unionOfArrayValidator,
      },
      {
        name: "union of primitives",
        input: `"a"`,
        expectedValue: "a",
        errorCode: "LiteralMismatch",
        validator: {
          type: "union",
          value: [validators.number, validators.boolean],
        },
      },
      {
        name: "record keys are not ids",
        input: `{a: 1}`,
        expectedValue: { a: 1 },
        errorCode: "RecordKeysMismatch",
        validator: {
          type: "record",
          keys: { type: "id", tableName: "abc" },
          values: {
            fieldType: validators.number,
            optional: false,
          },
        },
      },
      {
        name: "record values do not match",
        input: `{a: 1}`,
        expectedValue: { a: 1 },
        errorCode: "LiteralMismatch",
        validator: {
          type: "record",
          keys: { type: "string" },
          values: {
            fieldType: validators.string,
            optional: false,
          },
        },
      },
      {
        name: "union with records inside",
        input: `{a: false}`,
        expectedValue: { a: false },
        errorCode: "UnionMismatch",
        validator: {
          type: "union",
          value: [
            {
              type: "record",
              keys: { type: "string" },
              values: {
                fieldType: validators.string,
                optional: false,
              },
            },
            {
              type: "record",
              keys: { type: "string" },
              values: {
                fieldType: validators.number,
                optional: false,
              },
            },
          ],
        },
      },
      {
        name: "record with invalid object inside",
        input: `{ key: {a: 1} }`,
        expectedValue: { key: { a: 1 } },
        errorCode: "LiteralMismatch",
        validator: {
          type: "record",
          keys: { type: "string" },
          values: {
            fieldType: {
              type: "object",
              value: {
                a: { fieldType: validators.string, optional: false },
              },
            },
            optional: false,
          },
        },
      },
    ];

    // TODO: Test that multiple errors can be produced.

    test.each(testCases)(
      "$name",
      ({ input, errorCode, validator, expectedValue }) => {
        const { value, errors } = walkAst(input, {
          validator,
          mode: "editField",
        });

        expect(errors).toEqual([
          expect.objectContaining({
            code: errorCode,
          }),
        ]);
        if (expectedValue !== undefined) {
          expect(value).toEqual(expectedValue);
        }
      },
    );
  });
});

describe("validateConvexFieldName", () => {
  test.each([
    {
      name: "starts with $",
      input: "$a",
      isTopLevel: false,
      expected: "Field cannot start with a '$'",
    },
    {
      name: "starts with _",
      input: "_a",
      isTopLevel: true,
      expected: "Field is top-level and cannot start with an underscore.",
    },
    {
      name: "nested starts with _",
      input: "_b",
      isTopLevel: false,
      expected: undefined,
    },
    {
      name: "contains non-ascii characters",
      input: "a\u0000",
      isTopLevel: false,
      expected: "Field must only contain non-control ASCII characters.",
    },
  ])(
    "returns the correct error message for a $name",
    ({ input, isTopLevel, expected }) => {
      const returned = validateConvexFieldName(input, "Field", isTopLevel);
      expect(returned).toEqual(expected);
    },
  );
});
