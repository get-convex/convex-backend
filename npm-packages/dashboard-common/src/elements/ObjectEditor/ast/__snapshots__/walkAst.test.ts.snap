// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
[
  [ConvexValidationError: Unsupported field name: field is top-level and cannot start with an underscore.],
]
`;

exports[`walkAst error cases -nan 1`] = `
[
  [ConvexValidationError: "-" must be followed by a number or bigint.],
]
`;

exports[`walkAst error cases Bytes with bad argument 1`] = `
[
  [ConvexValidationError: The Bytes constructor requires a valid base64 encoded string: Invalid string. Length must be a multiple of 4],
]
`;

exports[`walkAst error cases Bytes with no arguments 1`] = `
[
  [ConvexValidationError: The Bytes constructor requires exactly one argument.],
]
`;

exports[`walkAst error cases Bytes with too many arguments 1`] = `
[
  [ConvexValidationError: The Bytes constructor requires exactly one argument.],
]
`;

exports[`walkAst error cases CallExpression that isn't bytes 1`] = `
[
  [ConvexValidationError: Unsupported call expression: "foo".],
]
`;

exports[`walkAst error cases Id class 1`] = `
[
  [ConvexValidationError: The \`Id\` class is no longer supported. Use an ID string instead.],
]
`;

exports[`walkAst error cases array with empty value 1`] = `
[
  [ConvexValidationError: Arrays must not have empty elements],
]
`;

exports[`walkAst error cases identifier 1`] = `
[
  [ConvexValidationError: \`a\` is not a valid Convex value],
]
`;

exports[`walkAst error cases identifier in object 1`] = `
[
  [ConvexValidationError: \`b\` is not a valid Convex value],
]
`;

exports[`walkAst error cases invalid unary expression 1`] = `
[
  [ConvexValidationError: Unsupported UnaryExpression: "+"],
]
`;

exports[`walkAst error cases key starting with $ 1`] = `
[
  [ConvexValidationError: Unsupported field name: field cannot start with a '$'],
]
`;

exports[`walkAst error cases negative object 1`] = `
[
  [ConvexValidationError: "-" must be followed by a number or bigint.],
]
`;

exports[`walkAst error cases negative string 1`] = `
[
  [ConvexValidationError: "-" must be followed by a number or bigint.],
]
`;

exports[`walkAst error cases nested undefined in object even if top level is allowed 1`] = `
[
  [ConvexValidationError: \`undefined\` is not a valid Convex value],
]
`;

exports[`walkAst error cases non-identifier key 1`] = `[]`;

exports[`walkAst error cases regex 1`] = `
[
  [ConvexValidationError: Unsupported syntax: RegExp],
]
`;

exports[`walkAst error cases template literal with expression 1`] = `
[
  [ConvexValidationError: Unsupported template literal: expressions are not supported.],
]
`;

exports[`walkAst error cases top level undefined 1`] = `
[
  [ConvexValidationError: \`undefined\` is not a valid Convex value],
]
`;

exports[`walkAst error cases top level undefined in object 1`] = `
[
  [ConvexValidationError: \`undefined\` is not a valid Convex value],
]
`;

exports[`walkAst error cases top level undefined in object 2`] = `
[
  [ConvexValidationError: \`undefined\` is not a valid Convex value],
]
`;

exports[`walkAst error cases unsupported constructor 1`] = `
[
  [ConvexValidationError: Unsupported constructor: "User".],
]
`;
